File: apps/android/app/src/main/java/chat/vostok/android/core/network/ApiClient.kt
package chat.vostok.android.core.network

import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory

class ApiClient(
    baseUrl: String,
    tokenProvider: () -> String?
) {
    private val service: ApiEndpoints

    init {
        val logger = HttpLoggingInterceptor().apply { level = HttpLoggingInterceptor.Level.BASIC }
        val client = OkHttpClient.Builder()
            .addInterceptor(AuthInterceptor(tokenProvider))
            .addInterceptor(logger)
            .build()

        val moshi = Moshi.Builder()
            .addLast(KotlinJsonAdapterFactory())
            .build()

        service = Retrofit.Builder()
            .baseUrl(baseUrl)
            .client(client)
            .addConverterFactory(MoshiConverterFactory.create(moshi))
            .build()
            .create(ApiEndpoints::class.java)
    }

    suspend fun health(): HealthResponse = service.health()

    suspend fun bootstrap(): BootstrapResponse = service.bootstrap()

    suspend fun register(request: RegisterRequest): RegistrationResponse = service.register(request)

    suspend fun challenge(deviceId: String): ChallengeResponse =
        service.challenge(ChallengeRequest(deviceId = deviceId))

    suspend fun verify(
        deviceId: String,
        challengeId: String,
        signatureBase64: String
    ): VerifyResponse =
        service.verify(
            VerifyChallengeRequest(
                deviceId = deviceId,
                challengeId = challengeId,
                signature = signatureBase64
            )
        )

    suspend fun me(): MeResponse = service.me()

    suspend fun publishPrekeys(request: PublishPrekeysRequest): PublishPrekeysResponse =
        service.publishPrekeys(request)

    suspend fun userPrekeys(username: String): UserPrekeysResponse = service.userPrekeys(username)

    suspend fun chats(): ChatsResponse = service.chats()

    suspend fun createDirectChat(username: String): ChatEnvelopeResponse =
        service.createDirectChat(CreateDirectChatRequest(username = username))

    suspend fun createGroupChat(title: String, members: List<String>): ChatEnvelopeResponse =
        service.createGroupChat(CreateGroupChatRequest(title = title, members = members))

    suspend fun renameGroupChat(chatId: String, title: String): ChatEnvelopeResponse =
        service.renameGroupChat(chatId, RenameGroupChatRequest(title = title))

    suspend fun groupMembers(chatId: String): GroupMembersResponse = service.groupMembers(chatId)

    suspend fun updateGroupMemberRole(chatId: String, userId: String, role: String): GroupMemberEnvelopeResponse =
        service.updateGroupMemberRole(chatId, userId, UpdateGroupMemberRoleRequest(role = role))

    suspend fun removeGroupMember(chatId: String, userId: String): GroupMemberEnvelopeResponse =
        service.removeGroupMember(chatId, userId)

    suspend fun safetyNumbers(chatId: String): SafetyNumbersResponse = service.safetyNumbers(chatId)

    suspend fun verifySafetyNumber(chatId: String, peerDeviceId: String): SafetyNumberEnvelopeResponse =
        service.verifySafetyNumber(chatId, peerDeviceId)

    suspend fun messages(chatId: String): MessagesResponse = service.messages(chatId)

    suspend fun recipientDevices(chatId: String): RecipientDevicesResponse =
        service.recipientDevices(chatId)

    suspend fun sessionBootstrap(chatId: String, peerDeviceId: String): SessionBootstrapResponse =
        service.sessionBootstrap(chatId, SessionBootstrapRequest(peerDeviceId))

    suspend fun sessionRekey(chatId: String, peerDeviceId: String): SessionBootstrapResponse =
        service.sessionRekey(chatId, SessionBootstrapRequest(peerDeviceId))

    suspend fun createMessage(chatId: String, request: CreateMessageRequest): MessageResponse =
        service.createMessage(chatId, request)

    suspend fun updateMessage(chatId: String, messageId: String, request: UpdateMessageRequest): MessageResponse =
        service.updateMessage(chatId, messageId, request)

    suspend fun deleteMessage(chatId: String, messageId: String): MessageResponse =
        service.deleteMessage(chatId, messageId)

    suspend fun togglePin(chatId: String, messageId: String): MessageResponse =
        service.togglePin(chatId, messageId)

    suspend fun toggleReaction(chatId: String, messageId: String, reactionKey: String): MessageResponse =
        service.toggleReaction(chatId, messageId, ToggleReactionRequest(reactionKey = reactionKey))

    suspend fun createMediaUpload(
        mediaKind: String,
        filename: String,
        contentType: String,
        declaredByteSize: Int,
        expectedPartCount: Int?
    ): MediaUploadEnvelopeResponse =
        service.createMediaUpload(
            CreateUploadRequest(
                mediaKind = mediaKind,
                filename = filename,
                contentType = contentType,
                declaredByteSize = declaredByteSize,
                expectedPartCount = expectedPartCount
            )
        )

    suspend fun mediaUploadStatus(uploadId: String): MediaUploadEnvelopeResponse =
        service.mediaUploadStatus(uploadId)

    suspend fun uploadMediaPart(
        uploadId: String,
        chunkBase64: String,
        partIndex: Int?,
        partCount: Int?
    ): MediaUploadEnvelopeResponse =
        service.uploadMediaPart(
            uploadId,
            UploadPartRequest(
                chunk = chunkBase64,
                partIndex = partIndex,
                partCount = partCount
            )
        )

    suspend fun completeMediaUpload(
        uploadId: String,
        ciphertextSha256: String?
    ): MediaUploadEnvelopeResponse =
        service.completeMediaUpload(uploadId, CompleteUploadRequest(ciphertextSha256 = ciphertextSha256))

    suspend fun mediaById(uploadId: String): MediaUploadEnvelopeResponse =
        service.mediaById(uploadId)

    suspend fun linkMetadata(url: String): LinkMetadataResponse =
        service.linkMetadata(LinkMetadataRequest(url = url))

    suspend fun devices(): DevicesResponse = service.devices()

    suspend fun revokeDevice(deviceId: String): DeviceEnvelopeResponse = service.revokeDevice(deviceId)

    suspend fun activeCall(chatId: String): ActiveCallResponse = service.activeCall(chatId)

    suspend fun createCall(chatId: String, mode: String): CallEnvelopeResponse =
        service.createCall(chatId, CreateCallRequest(mode = mode))

    suspend fun callState(callId: String): CallStateResponse = service.callState(callId)

    suspend fun joinCall(
        callId: String,
        trackKind: String,
        e2eeCapable: Boolean? = null,
        e2eeAlgorithm: String? = null,
        e2eeKeyEpoch: Int? = null
    ): JoinCallResponse =
        service.joinCall(
            callId,
            JoinCallRequest(
                trackKind = trackKind,
                e2eeCapable = e2eeCapable,
                e2eeAlgorithm = e2eeAlgorithm,
                e2eeKeyEpoch = e2eeKeyEpoch
            )
        )

    suspend fun callSignals(callId: String): SignalsResponse = service.callSignals(callId)

    suspend fun emitCallSignal(
        callId: String,
        signalType: String,
        payload: String,
        targetDeviceId: String? = null
    ): SignalEnvelopeResponse =
        service.emitCallSignal(
            callId,
            EmitSignalRequest(
                signalType = signalType,
                payload = payload,
                targetDeviceId = targetDeviceId
            )
        )

    suspend fun callKeys(callId: String): CallKeysResponse = service.callKeys(callId)

    suspend fun rotateCallKeys(
        callId: String,
        keyEpoch: Int,
        algorithm: String,
        wrappedKeys: Map<String, String>
    ): CallKeysResponse =
        service.rotateCallKeys(
            callId,
            RotateCallKeysRequest(
                keyEpoch = keyEpoch,
                algorithm = algorithm,
                wrappedKeys = wrappedKeys
            )
        )

    suspend fun provisionWebRtcEndpoint(callId: String): WebRtcEndpointResponse =
        service.provisionWebRtcEndpoint(callId)

    suspend fun webRtcEndpointState(callId: String): WebRtcEndpointResponse =
        service.webRtcEndpointState(callId)

    suspend fun pollWebRtcEndpoint(callId: String): WebRtcPollResponse =
        service.pollWebRtcEndpoint(callId)

    suspend fun leaveCall(callId: String): JoinCallResponse = service.leaveCall(callId)

    suspend fun endCall(callId: String): CallEnvelopeResponse = service.endCall(callId)

    suspend fun turnCredentials(ttlSeconds: Int = 3600): TurnCredentialsResponse =
        service.turnCredentials(TurnCredentialsRequest(ttlSeconds = ttlSeconds))
}

File: apps/ios/Vostok/Core/Networking/APIClient.swift
import Foundation

final class APIClient: VostokAPIClientProtocol {
    private let baseURL: URL
    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    init(baseURL: URL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
        self.decoder = JSONDecoder()
        self.encoder = JSONEncoder()
    }

    func health() async throws -> HealthResponse { try await request(path: "/health", method: "GET") }
    func apiHealth() async throws -> HealthResponse { try await request(path: "/api/v1/health", method: "GET") }
    func bootstrap() async throws -> BootstrapResponse { try await request(path: "/api/v1/bootstrap", method: "GET") }
    func ingestFederationDelivery(request payload: FederationDeliveryRequest) async throws -> FederationDeliveryResponse {
        try await request(path: "/api/v1/federation/deliveries", method: "POST", body: payload)
    }

    func acceptFederationPeerInvite(request payload: FederationPeerInviteAcceptRequest) async throws -> FederationPeerResponse {
        try await request(path: "/api/v1/federation/peers/accept", method: "POST", body: payload)
    }

    func register(request payload: RegisterRequest) async throws -> RegistrationResponse {
        try await request(path: "/api/v1/register", method: "POST", body: payload)
    }

    func challenge(deviceID: String) async throws -> ChallengeResponse {
        try await request(path: "/api/v1/auth/challenge", method: "POST", body: ["device_id": deviceID])
    }

    func verify(request payload: VerifyRequest) async throws -> VerifyResponse {
        try await request(path: "/api/v1/auth/verify", method: "POST", body: payload)
    }

    func me(token: String) async throws -> MeResponse { try await authed(path: "/api/v1/me", method: "GET", token: token) }

    func linkDevice(token: String, request: LinkDeviceRequest) async throws -> DeviceLinkResponse {
        try await authed(path: "/api/v1/devices/link", method: "POST", token: token, body: request)
    }

    func devices(token: String) async throws -> DevicesResponse {
        try await authed(path: "/api/v1/devices", method: "GET", token: token)
    }

    func revokeDevice(token: String, deviceID: String) async throws -> DevicesResponse {
        try await authed(path: "/api/v1/devices/\(deviceID)/revoke", method: "POST", token: token, body: EmptyBody())
    }

    func publishPrekeys(token: String, request: PublishPrekeysRequest) async throws -> PrekeyPublishResponse {
        try await authed(path: "/api/v1/devices/prekeys", method: "POST", token: token, body: request)
    }

    func prekeys(token: String, username: String) async throws -> PrekeyLookupResponse {
        try await authed(path: "/api/v1/users/\(username)/devices/prekeys", method: "GET", token: token)
    }

    func chats(token: String) async throws -> ChatsResponse { try await authed(path: "/api/v1/chats", method: "GET", token: token) }

    func createDirectChat(token: String, username: String) async throws -> ChatResponse {
        try await authed(path: "/api/v1/chats/direct", method: "POST", token: token, body: ["username": username])
    }

    func createGroup(token: String, request: CreateGroupRequest) async throws -> ChatResponse {
        try await authed(path: "/api/v1/chats/group", method: "POST", token: token, body: request)
    }

    func updateGroup(token: String, chatID: String, title: String) async throws -> ChatResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/group", method: "PATCH", token: token, body: ["title": title])
    }

    func groupMembers(token: String, chatID: String) async throws -> GroupMembersResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/members", method: "GET", token: token)
    }

    func updateGroupMember(token: String, chatID: String, userID: String, request: UpdateGroupMemberRequest) async throws -> GroupMemberResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/members/\(userID)", method: "PATCH", token: token, body: request)
    }

    func removeGroupMember(token: String, chatID: String, userID: String) async throws -> GroupMemberResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/members/\(userID)/remove", method: "POST", token: token, body: EmptyBody())
    }

    func recipientDevices(token: String, chatID: String) async throws -> RecipientDevicesResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/recipient-devices", method: "GET", token: token)
    }

    func messages(token: String, chatID: String) async throws -> MessagesResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/messages", method: "GET", token: token)
    }

    func createMessage(token: String, chatID: String, request: CreateMessageRequest) async throws -> MessageResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/messages", method: "POST", token: token, body: request)
    }

    func editMessage(token: String, chatID: String, messageID: String, request: EditMessageRequest) async throws -> MessageResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/messages/\(messageID)", method: "PATCH", token: token, body: request)
    }

    func deleteMessage(token: String, chatID: String, messageID: String) async throws -> MessageResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/messages/\(messageID)/delete", method: "POST", token: token, body: EmptyBody())
    }

    func togglePin(token: String, chatID: String, messageID: String) async throws -> MessageResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/messages/\(messageID)/pin", method: "POST", token: token, body: EmptyBody())
    }

    func toggleReaction(token: String, chatID: String, messageID: String, reactionKey: String) async throws -> MessageResponse {
        try await authed(
            path: "/api/v1/chats/\(chatID)/messages/\(messageID)/reactions",
            method: "POST",
            token: token,
            body: ["reaction_key": reactionKey]
        )
    }

    func senderKeys(token: String, chatID: String) async throws -> SenderKeysResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/sender-keys", method: "GET", token: token)
    }

    func distributeSenderKeys(token: String, chatID: String, request: DistributeSenderKeysRequest) async throws -> SenderKeysResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/sender-keys", method: "POST", token: token, body: request)
    }

    func sessionBootstrap(token: String, chatID: String, request: SessionBootstrapRequest) async throws -> SessionBootstrapResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/session-bootstrap", method: "POST", token: token, body: request)
    }

    func sessionRekey(token: String, chatID: String, request: SessionRekeyRequest) async throws -> SessionBootstrapResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/session-rekey", method: "POST", token: token, body: request)
    }

    func safetyNumbers(token: String, chatID: String) async throws -> SafetyNumbersResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/safety-numbers", method: "GET", token: token)
    }

    func verifySafetyNumber(token: String, chatID: String, peerDeviceID: String) async throws -> SafetyNumberDTO {
        struct Wrap: Codable { let safetyNumber: SafetyNumberDTO; enum CodingKeys: String, CodingKey { case safetyNumber = "safety_number" } }
        let wrap: Wrap = try await authed(
            path: "/api/v1/chats/\(chatID)/safety-numbers/\(peerDeviceID)/verify",
            method: "POST",
            token: token,
            body: EmptyBody()
        )
        return wrap.safetyNumber
    }

    func createUpload(token: String, request: CreateUploadRequest) async throws -> UploadResponse {
        try await authed(path: "/api/v1/media/uploads", method: "POST", token: token, body: request)
    }

    func uploadPart(token: String, id: String, request: UploadPartRequest) async throws -> UploadResponse {
        try await authed(path: "/api/v1/media/uploads/\(id)/part", method: "PATCH", token: token, body: request)
    }

    func uploadStatus(token: String, id: String) async throws -> UploadResponse {
        try await authed(path: "/api/v1/media/uploads/\(id)", method: "GET", token: token)
    }

    func completeUpload(token: String, id: String, request: CompleteUploadRequest) async throws -> UploadResponse {
        try await authed(path: "/api/v1/media/uploads/\(id)/complete", method: "POST", token: token, body: request)
    }

    func media(token: String, id: String) async throws -> UploadResponse {
        try await authed(path: "/api/v1/media/\(id)", method: "GET", token: token)
    }

    func linkMetadata(token: String, request: LinkMetadataRequest) async throws -> LinkMetadataResponse {
        try await authed(path: "/api/v1/media/link-metadata", method: "POST", token: token, body: request)
    }

    func adminOverview(token: String) async throws -> AdminOverviewResponse {
        try await authed(path: "/api/v1/admin/overview", method: "GET", token: token)
    }

    func federationPeers(token: String) async throws -> FederationPeersResponse {
        try await authed(path: "/api/v1/admin/federation/peers", method: "GET", token: token)
    }

    func createFederationPeer(token: String, request: CreateFederationPeerRequest) async throws -> FederationPeerResponse {
        try await authed(path: "/api/v1/admin/federation/peers", method: "POST", token: token, body: request)
    }

    func federationDeliveries(token: String) async throws -> FederationDeliveriesResponse {
        try await authed(path: "/api/v1/admin/federation/deliveries", method: "GET", token: token)
    }

    func createFederationDelivery(
        token: String,
        peerID: String,
        request: CreateFederationDeliveryRequest
    ) async throws -> FederationDeliveryResponse {
        try await authed(path: "/api/v1/admin/federation/peers/\(peerID)/deliveries", method: "POST", token: token, body: request)
    }

    func attemptFederationDelivery(
        token: String,
        jobID: String,
        request: AttemptFederationDeliveryRequest
    ) async throws -> FederationDeliveryResponse {
        try await authed(path: "/api/v1/admin/federation/deliveries/\(jobID)/attempt", method: "POST", token: token, body: request)
    }

    func updateFederationPeerStatus(
        token: String,
        peerID: String,
        request: UpdateFederationPeerStatusRequest
    ) async throws -> FederationPeerResponse {
        try await authed(path: "/api/v1/admin/federation/peers/\(peerID)/status", method: "POST", token: token, body: request)
    }

    func federationPeerHeartbeat(token: String, peerID: String) async throws -> FederationPeerResponse {
        try await authed(path: "/api/v1/admin/federation/peers/\(peerID)/heartbeat", method: "POST", token: token, body: EmptyBody())
    }

    func createFederationPeerInvite(token: String, peerID: String) async throws -> FederationPeerInviteResponse {
        try await authed(path: "/api/v1/admin/federation/peers/\(peerID)/invite", method: "POST", token: token, body: EmptyBody())
    }

    func createCall(token: String, chatID: String, mode: String) async throws -> CallResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/calls", method: "POST", token: token, body: ["mode": mode])
    }

    func activeCall(token: String, chatID: String) async throws -> CallResponse {
        try await authed(path: "/api/v1/chats/\(chatID)/calls/active", method: "GET", token: token)
    }

    func callState(token: String, callID: String) async throws -> CallResponse {
        try await authed(path: "/api/v1/calls/\(callID)", method: "GET", token: token)
    }

    func joinCall(token: String, callID: String, request: JoinCallRequest) async throws -> CallResponse {
        try await authed(path: "/api/v1/calls/\(callID)/join", method: "POST", token: token, body: request)
    }

    func callKeys(token: String, callID: String) async throws -> CallKeysResponse {
        try await authed(path: "/api/v1/calls/\(callID)/keys", method: "GET", token: token)
    }

    func rotateCallKeys(token: String, callID: String, request: RotateCallKeysRequest) async throws -> CallKeysResponse {
        try await authed(path: "/api/v1/calls/\(callID)/keys", method: "POST", token: token, body: request)
    }

    func provisionEndpoint(token: String, callID: String, request: ProvisionWebRTCEndpointRequest) async throws -> WebRTCEndpointResponse {
        try await authed(path: "/api/v1/calls/\(callID)/webrtc-endpoint", method: "POST", token: token, body: request)
    }

    func callSignals(token: String, callID: String) async throws -> CallSignalsResponse {
        try await authed(path: "/api/v1/calls/\(callID)/signals", method: "GET", token: token)
    }

    func emitSignal(token: String, callID: String, request: EmitSignalRequest) async throws -> CallSignalDTO {
        struct Wrap: Codable { let signal: CallSignalDTO }
        let wrap: Wrap = try await authed(path: "/api/v1/calls/\(callID)/signals", method: "POST", token: token, body: request)
        return wrap.signal
    }

    func endpointState(token: String, callID: String) async throws -> WebRTCEndpointResponse {
        try await authed(path: "/api/v1/calls/\(callID)/webrtc-endpoint", method: "GET", token: token)
    }

    func pushEndpointMediaEvent(token: String, callID: String, request: PushWebRTCMediaEventRequest) async throws -> WebRTCEndpointResponse {
        try await authed(path: "/api/v1/calls/\(callID)/webrtc-endpoint/media-events", method: "POST", token: token, body: request)
    }

    func pollEndpoint(token: String, callID: String) async throws -> WebRTCEndpointResponse {
        try await authed(path: "/api/v1/calls/\(callID)/webrtc-endpoint/poll", method: "POST", token: token, body: EmptyBody())
    }

    func leaveCall(token: String, callID: String) async throws -> CallResponse {
        try await authed(path: "/api/v1/calls/\(callID)/leave", method: "POST", token: token, body: EmptyBody())
    }

    func endCall(token: String, callID: String) async throws -> CallResponse {
        try await authed(path: "/api/v1/calls/\(callID)/end", method: "POST", token: token, body: EmptyBody())
    }

    func turnCredentials(token: String) async throws -> TurnCredentialsResponse {
        try await authed(path: "/api/v1/calls/turn-credentials", method: "POST", token: token, body: EmptyBody())
    }

    private struct EmptyBody: Codable {}

    private func request<Response: Decodable>(path: String, method: String) async throws -> Response {
        try await request(path: path, method: method, body: Optional<EmptyBody>.none, token: nil)
    }

    private func request<Response: Decodable, Body: Encodable>(path: String, method: String, body: Body) async throws -> Response {
        try await request(path: path, method: method, body: Optional(body), token: nil)
    }

    private func authed<Response: Decodable>(path: String, method: String, token: String) async throws -> Response {
        try await request(path: path, method: method, body: Optional<EmptyBody>.none, token: token)
    }

    private func authed<Response: Decodable, Body: Encodable>(path: String, method: String, token: String, body: Body) async throws -> Response {
        try await request(path: path, method: method, body: Optional(body), token: token)
    }

    private func request<Response: Decodable, Body: Encodable>(path: String, method: String, body: Body?, token: String?) async throws -> Response {
        guard let url = URL(string: path, relativeTo: baseURL) else {
            throw VostokAPIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        if let body {
            request.httpBody = try encoder.encode(body)
        }

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                throw VostokAPIError.invalidResponse
            }

            guard (200...299).contains(http.statusCode) else {
                let envelope = try? decoder.decode(ErrorEnvelope.self, from: data)
                let message = envelope?.message ?? "Request failed with status \(http.statusCode)"
                if http.statusCode == 401 { throw VostokAPIError.unauthorized(message) }
                if http.statusCode == 404 { throw VostokAPIError.notFound(message) }
                throw VostokAPIError.validation(message)
            }

            return try decoder.decode(Response.self, from: data)
        } catch let error as VostokAPIError {
            throw error
        } catch {
            throw VostokAPIError.transport(error)
        }
    }
}
